core/include/traccc/utils/particle.hpp

```diff
@@ -0,0 +1,98 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// traccc include(s).
+#include "traccc/definitions/primitives.hpp"
+#include "traccc/definitions/qualifiers.hpp"
+#include "traccc/edm/track_parameters.hpp"
+
+// detray include(s).
+#include "detray/definitions/pdg_particle.hpp"
+
+// System include(s).
+#include <stdexcept>
+
+namespace traccc {
+
+namespace detail {
+
+template <typename scalar_t>
+TRACCC_HOST_DEVICE inline detray::pdg_particle<scalar_t>
+particle_from_pdg_number(const std::int32_t pdg_num) {
+
+    if (pdg_num == 11) {
+        return detray::electron<scalar_t>();
+    } else if (pdg_num == -11) {
+        return detray::positron<scalar_t>();
+    } else if (pdg_num == 13) {
+        return detray::muon<scalar_t>();
+    } else if (pdg_num == -13) {
+        return detray::antimuon<scalar_t>();
+    } else if (pdg_num == 211) {
+        return detray::pion_plus<scalar_t>();
+    } else if (pdg_num == -211) {
+        return detray::pion_minus<scalar_t>();
+    }
+
+    return detray::muon<scalar_t>();
+}
+
+// Apply the charge operator to return the antimatter
+template <typename scalar_t>
+TRACCC_HOST_DEVICE inline detray::pdg_particle<scalar_t> charge_operator(
+    const detray::pdg_particle<scalar_t>& ptc) {
+
+    const auto pdg_num = ptc.pdg_num();
+
+    if (pdg_num == 11) {
+        return detray::positron<scalar_t>();
+    } else if (pdg_num == -11) {
+        return detray::electron<scalar_t>();
+    } else if (pdg_num == 13) {
+        return detray::antimuon<scalar_t>();
+    } else if (pdg_num == -13) {
+        return detray::muon<scalar_t>();
+    } else if (pdg_num == 211) {
+        return detray::pion_minus<scalar_t>();
+    } else if (pdg_num == -211) {
+        return detray::pion_plus<scalar_t>();
+    }
+
+    return detray::muon<scalar_t>();
+}
+
+// Return the consistent particle type based on the particle hypothesis and the
+// charge of the track parameters
+template <typename scalar_t>
+TRACCC_HOST_DEVICE inline detray::pdg_particle<scalar_t> consistent_particle(
+    const detray::pdg_particle<scalar_t>& ptc_hypothesis,
+    const bound_track_parameters& params) {
+
+    if (ptc_hypothesis.charge() * params.qop() > 0.f) {
```

Comment:
This PR does not setup the hypothesis for the individual tracks but it is event-wise.

Reply:
> Would we not actually need to monitor updates to the curvature of the track with regard to the B-field or am I confused?

I don't get this question

Reply:
Ah, Ok, because it seemed to call ```consistent_particle``` for every new step of the CKF and the material interaction, but I probably just don't know the code well enough

Reply:
If we don't like this, we may have to put the particle hypothesis into the bound track parameters, but lets not do that.